### PR TITLE
feat: add maximum age to replication queues

### LINKED ISSUE #23109: Maximum age config for data in the replications queue
__Proposal:__
Add a max age to replications data

__Current behavior:__
Currently it is possible for data to become 'stuck' in a replications queue if the remote node never accepts it. This blocks everything else in the queue as well.

__Desired behavior:__
the data should be able to (configurably) age out

__Alternatives considered:__
 * Some server-side code (e.g. in `influxd inspect`) to dump the writes from a queue and clear the queue.
 * an API to manually dump 'stuck' data

__Use case:__
Using replications with a remote InfluxDB that doesn't like the replicated data for some reason.

### PR DESCRIPTION
**Pull request by DStrand1**
Closes #23109 

Introduces a maximum age for replications so that data can be timed out and dropped (configurable).
A value of `0` means that there is no time-out.

### CODE REVIEW

File: replications/remotewrite/writer.go [hunk 1]
```diff
@@ -85,91 +86,83 @@ func NewWriter(replicationID platform.ID, store HttpConfigStore, metrics *metric
...
+	if err != nil {
+		return w.backoff(attempts), true, err
+	}
 
-	for {
+	res, postWriteErr := PostWrite(ctx, conf, data, w.clientTimeout)
+	res, msg, ok := normalizeResponse(res, postWriteErr)
+	if !ok {
+		// bail out
+		return w.backoff(attempts), true, postWriteErr
```
Comment (lesam) [inline replications/remotewrite/writer.go:117]: I think this code is right, but `normalizeResponse` says the bool is for if the error is retryable. I think the comment on `normalizeResponse` is lying here?

File: replications/remotewrite/writer.go [hunk 2]
```diff
@@ -85,91 +86,83 @@ func NewWriter(replicationID platform.ID, store HttpConfigStore, metrics *metric
...
-		case http.StatusTooManyRequests:
-			headerTime := w.waitTimeFromHeader(res)
-			if headerTime != 0 {
-				waitTime = headerTime
-				hasSetWaitTime = true
-			}
+	switch res.StatusCode {
+	case http.StatusBadRequest:
+		if conf.DropNonRetryableData {
+			w.logger.Debug("dropped data", zap.Int("bytes", len(data)))
```
Comment (lesam) [inline replications/remotewrite/writer.go:146]: * this should be `warn` level debugging I think- dropped data is serious.
* maybe we can print the error message from the body in the log as well?

**Review by lesam — COMMENTED** [review]
Looks good! Exciting

File: replications/internal/queue_management.go [hunk 3]
```diff
@@ -403,9 +434,16 @@ func (qm *durableQueueManager) EnqueueData(replicationID platform.ID, data []byt
 	return nil
 }
 
-func (qm *durableQueueManager) newReplicationQueue(id platform.ID, orgID platform.ID, localBucketID platform.ID, queue *durablequeue.Queue) *replicationQueue {
+func (qm *durableQueueManager) newReplicationQueue(id platform.ID, orgID platform.ID, localBucketID platform.ID, queue *durablequeue.Queue, maxAge int64) *replicationQueue {
 	logger := qm.logger.With(zap.String("replication_id", id.String()))
 	done := make(chan struct{})
+	// check for max age minimum
+	var maxAgeTime time.Duration
+	if maxAge < 0 {
```
[the remark this answers is not in the document]
Comment (DStrand1) [reply replications/internal/queue_management.go:442]: I left `0` as a special value for "no maximum age"

File: replications/remotewrite/writer_test.go [hunk 4]
```diff
@@ -226,21 +255,19 @@ func TestWrite_Metrics(t *testing.T) {
...
 	}{
 		{
-			name: "server errors",
-			status: func(i int) int {
-				arr := []int{http.StatusTeapot, http.StatusTeapot, http.StatusTeapot, http.StatusNoContent}
-				return arr[i]
-			},
-			data: []byte{},
+			name:        "server errors",
+			status:      constantStatus(http.StatusTeapot),
```
Comment (lesam) [inline replications/remotewrite/writer_test.go:265]: This is easier to follow I think

**Review by lesam — APPROVED** [review]

**Pull request merged**